### ISSUE TITLE
Bug in Core with advanced cmd option "invertBinary"

### ISSUE THREAD

**Issue by BadWolf42** [jeedom/core#2068]
Hello Team,

As per [this Community thread](https://community.jeedom.com/t/bug-avec-les-binaires-jmqtt-et-les-fonctions-inverser-les-valeurs-binaires-repeter-les-valeurs-identiques/91692), an issue has been found with cmd **configuration option** `invertBinary` here:
![image](https://user-images.githubusercontent.com/8396512/193443375-fd897c52-4788-4faa-8e6c-3a92de8a4b8e.png)
(So it is not the **display option** `invertBinary`)

If **configuration** key `invertBinary` = 1 and `repeatEventManagement` = 'never', 
Then binary value is only updated on repetition, so value is stuck.

As per my investigation, I was able to reproduce using jMQTT and Virtual plugin (see bellow).
The root cause is in [`eqLogic::checkAndUpdateCmd()`](https://github.com/jeedom/core/blob/ae8266b7567eac4592b6fe7fe22ddf6d610667d1/core/class/eqLogic.class.php#L635-L663) method:
- `$oldValue = $cmd->execCmd()`
retrieves the last (flipped) value in Jeedom -> OK
- `$oldValue !== $cmd->formatValue($_value)`
oldValue is comparted to current real value from device (unflipped),
but `cmd::formatValue()` does not take in account `invertBinary` config,
so this condition is NOT RIGHT when using `invertBinary`.
- `$cmd->event($_value, $_updateTime);`
updates the value in cmd/db/cache,
it flips binary values according `invertBinary`, but [too late](https://github.com/jeedom/core/blob/ae8266b7567eac4592b6fe7fe22ddf6d610667d1/core/class/cmd.class.php#L1701-L1703).
&nbsp;

Tell me if you want me to PR on this issue, if so, which way you want to correct:
- only `invertBinary` check in `eqLogic::checkAndUpdateCmd()`?
- move `invertBinary` handling from `cmd::event()` to `cmd::formatValue()`?
- move `invertBinary` handling from `cmd::event()` to `eqLogic::checkAndUpdateCmd()`?
- remove completely `invertBinary` (if this issue only appears now, it is not used)?
- something else?

&nbsp;

Issue reproduction on jMQTT with json payload:
```
# TOPIC val (json) - Inverser les valeurs binaires DESACTIVE

# ON  -> FALSE -> 0 => OK
0006|[2022-10-01 18:18:31][INFO] : Cmd #[Home][Topic91692][On]# -> '{"Contact":false}' sur le topic 91692/val (qos=1, retain=0)
0008|[2022-10-01 18:18:32][DEBUG] : Payload '{"Contact":false}' reçu sur le Topic '91692/val'
0009|[2022-10-01 18:18:32][DEBUG] : Cmd #[Home][Topic91692][Etat]# est de type info JSON : ignorée
0010|[2022-10-01 18:18:32][INFO] : Cmd #[Home][Topic91692][val]# <- {"Contact":false}
0011|[2022-10-01 18:18:32][INFO] : Cmd #[Home][Topic91692][Etat]# <- 0

# OFF -> TRUE -> 1 => OK
0012|[2022-10-01 18:18:36][INFO] : Cmd #[Home][Topic91692][Off]# -> '{"Contact":true}' sur le topic 91692/val (qos=1, retain=0)
0014|[2022-10-01 18:18:36][DEBUG] : Payload '{"Contact":true}' reçu sur le Topic '91692/val'
0015|[2022-10-01 18:18:36][DEBUG] : Cmd #[Home][Topic91692][Etat]# est de type info JSON : ignorée
0016|[2022-10-01 18:18:36][INFO] : Cmd #[Home][Topic91692][val]# <- {"Contact":true}
0017|[2022-10-01 18:18:36][INFO] : Cmd #[Home][Topic91692][Etat]# <- 1

# Inverser les valeurs binaires ACTIVE (#[Home][Topic91692][Etat]# = 1)

# ON  -> FALSE -> 1 => OK
0036|[2022-10-01 18:20:29][INFO] : Cmd #[Home][Topic91692][On]# -> '{"Contact":false}' sur le topic 91692/val (qos=1, retain=0)
0038|[2022-10-01 18:20:29][DEBUG] : Payload '{"Contact":false}' reçu sur le Topic '91692/val'
0039|[2022-10-01 18:20:29][DEBUG] : Cmd #[Home][Topic91692][Etat]# est de type info JSON : ignorée
0040|[2022-10-01 18:20:29][INFO] : Cmd #[Home][Topic91692][val]# <- {"Contact":false}
0041|[2022-10-01 18:20:29][INFO] : Cmd #[Home][Topic91692][Etat]# <- 1

# OFF -> TRUE -> 1 => NOT OK
0048|[2022-10-01 18:20:34][INFO] : Cmd #[Home][Topic91692][Off]# -> '{"Contact":true}' sur le topic 91692/val (qos=1, retain=0)
0050|[2022-10-01 18:20:34][DEBUG] : Payload '{"Contact":true}' reçu sur le Topic '91692/val'
0051|[2022-10-01 18:20:34][DEBUG] : Cmd #[Home][Topic91692][Etat]# est de type info JSON : ignorée
0052|[2022-10-01 18:20:34][INFO] : Cmd #[Home][Topic91692][val]# <- {"Contact":true}
0053|[2022-10-01 18:20:34][INFO] : Cmd #[Home][Topic91692][Etat]# <- 1
```

Issue reproduction on jMQTT with a binary info cmd:
```
# TOPIC val2 (pas json) - Inverser les valeurs binaires DESACTIVE

# ON2  -> FALSE -> 0 => OK
0111|[2022-10-01 18:25:07][INFO] : Cmd #[Home][Topic91692][On2]# -> 'false' sur le topic 91692/val2 (qos=1, retain=0)
0113|[2022-10-01 18:25:07][DEBUG] : Payload 'false' reçu sur le Topic '91692/val2'
0114|[2022-10-01 18:25:07][INFO] : Cmd #[Home][Topic91692][Etat2]# <- 0

# OFF2 -> TRUE -> 1 => OK
0115|[2022-10-01 18:25:09][INFO] : Cmd #[Home][Topic91692][Off2]# -> 'true' sur le topic 91692/val2 (qos=1, retain=0)
0117|[2022-10-01 18:25:09][DEBUG] : Payload 'true' reçu sur le Topic '91692/val2'
0118|[2022-10-01 18:25:09][INFO] : Cmd #[Home][Topic91692][Etat2]# <- 1

# Inverser les valeurs binaires ACTIVE (#[Home][Topic91692][Etat]# = 1)

# ON2  -> FALSE -> 1 => OK
0119|[2022-10-01 18:25:16][INFO] : Cmd #[Home][Topic91692][On2]# -> 'false' sur le topic 91692/val2 (qos=1, retain=0)
0121|[2022-10-01 18:25:16][DEBUG] : Payload 'false' reçu sur le Topic '91692/val2'
0122|[2022-10-01 18:25:16][INFO] : Cmd #[Home][Topic91692][Etat2]# <- 1

# OFF2 -> TRUE -> 1 => NOT OK
0123|[2022-10-01 18:25:21][INFO] : Cmd #[Home][Topic91692][Off2]# -> 'true' sur le topic 91692/val2 (qos=1, retain=0)
0125|[2022-10-01 18:25:21][DEBUG] : Payload 'true' reçu sur le Topic '91692/val2'
0126|[2022-10-01 18:25:21][INFO] : Cmd #[Home][Topic91692][Etat2]# <- 1
```
Config:
![image](https://user-images.githubusercontent.com/8396512/193443739-5b63f869-a128-420c-af3c-f218f84facc6.png)

Issue reproduction on Virtual with a binary info cmd:
```
# Inverser les valeurs binaires DESACTIVE
# ON  -> FALSE -> 0 => OK
0005|[2022-10-01 18:42:34]INFO : Exécution de la commande [Home][Virt91692][On] avec les paramètres {"user_login":"admin","user_id":"1"}
0006|[2022-10-01 18:42:34]INFO : Evènement sur la commande [Home][Virt91692][Etat] valeur : 0

# OFF -> TRUE -> 1 => OK
0007|[2022-10-01 18:42:38]INFO : Exécution de la commande [Home][Virt91692][Off] avec les paramètres {"user_login":"admin","user_id":"1"}
0008|[2022-10-01 18:42:38]INFO : Evènement sur la commande [Home][Virt91692][Etat] valeur : 1

# Inverser les valeurs binaires ACTIVE (#[Home][Topic91692][Etat]# = 1)
# ON  -> FALSE -> RIEN => NOT OK
0009|[2022-10-01 18:42:53]INFO : Exécution de la commande [Home][Virt91692][On] avec les paramètres {"user_login":"admin","user_id":"1"}

# OFF -> TRUE -> RIEN => NOT OK
0010|[2022-10-01 18:42:56]INFO : Exécution de la commande [Home][Virt91692][Off] avec les paramètres {"user_login":"admin","user_id":"1"}

# ON  -> FALSE -> RIEN => NOT OK
0011|[2022-10-01 18:42:59]INFO : Exécution de la commande [Home][Virt91692][On] avec les paramètres {"user_login":"admin","user_id":"1"}

# OFF -> TRUE -> RIEN => NOT OK
0018|[2022-10-01 18:43:01]INFO : Exécution de la commande [Home][Virt91692][Off] avec les paramètres {"user_login":"admin","user_id":"1"}
```
Config:
![image](https://user-images.githubusercontent.com/8396512/193443733-562add31-6d75-4417-96f6-960836dc38d9.png)
&nbsp;

Thanks,
Bad

**Comment by kwizer15** [comment]
try to change the `!==` by `!=` in `$oldValue !== $cmd->formatValue($_value)`. I think the values are not the same type…

**Comment by BadWolf42** [comment]
Hello,

As Jeedom 4.3 will be released on Monday, I would like to introduce a fix by then.

I suggested multiple ways, can you tell me which one should I implement?

Thanks,
Bad

**Comment by BadWolf42** [comment]
Fixed by PR #2075

Thank you
